### PR TITLE
Release v0.2.0

### PR DESCRIPTION
**Pull request by JoeZijunZhou**
- Release JetStream v0.2.0 with tag
- Add release github action (CI/CD) to auto release and publish on push with version tag
- Add JetStream MaxText TPU inference user guide

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,6 +4,301 @@
...
+weight_dtype=${WEIGHT_DTYPE} \
+per_device_batch_size=${PER_DEVICE_BATCH_SIZE} \
+quantization=${QUANTIZATION} \
+quantize_kvcache=${QUANTIZE_KVCACHE}
+```
+
+### Benchmarking Gemma-7b
+
+```bash
+# For Gemma 7b, plz use Gemma tokenizer (must use a tokenizer that matches your model; double check if you are using tokenizer.gemma) and ShareGPT dataset; add `--warmup-first` flag for your 1st run to warmup the server.
```
Comment (yeandy) [inline README.md:249]:
```suggestion
Instructions
- Download the ShareGPT dataset
- Make sure to use the Gemma tokenizer (tokenizer.gemma) when running Gemma 7b.
- Add `--warmup-first` flag for your 1st run to warmup the server

```bash
```

File: README.md [hunk 2]
```diff
@@ -4,6 +4,301 @@
...
+
+## Prerequisites: Prepare your GCP project and connect to Cloud TPU VM
+
+Follow the steps in [Manage TPU resources | Google Cloud](https://cloud.google.com/tpu/docs/managing-tpus-tpu-vm) to create a Cloud TPU VM (Recommend TPU type: `v5litepod-8`) and connect to the Cloud TPU VM.
+
+
+## Step 1: Download JetStream and the MaxText github repository
+
+```bash
+git clone -b jetstream-v0.2.0 https://github.com/google/maxtext.git
```
Comment (yeandy) [inline README.md:37]: Will this need to be updated with every jetstream release?

File: README.md [hunk 3]
```diff
@@ -4,6 +4,301 @@
...
+
+*   You can download a [Gemma checkpoint from Kaggle](https://www.kaggle.com/models/google/gemma/frameworks/maxText/variations/7b). 
+*   After downloading checkpoints, copy them to your GCS bucket at `$CHKPT_BUCKET`.
+    *   `gsutil -m cp -r ${YOUR_CKPT_PATH} ${CHKPT_BUCKET}`
+    *   Please refer to the [conversion script](https://github.com/google/JetStream/blob/main/jetstream/tools/maxtext/model_ckpt_conversion.sh) for an example of `$CHKPT_BUCKET`.
+*   Then, using the following command to convert the Gemma checkpoint into a MaxText compatible unscanned checkpoint.
+
+```bash
+# bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh ${MODEL} ${MODEL_VARIATION} ${CHKPT_BUCKET}
+bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh gemma 7b ${CHKPT_BUCKET}
```
Comment (yeandy) [inline README.md:68]:
```suggestion
# bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh ${MODEL} ${MODEL_VARIATION} ${CHKPT_BUCKET}

# For gemma-7b
bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh gemma 7b ${CHKPT_BUCKET}
```

File: README.md [hunk 4]
```diff
@@ -4,6 +4,301 @@
...
+
+
+## Step4: Run the JetStream MaxText server
+
+
+### Create model config environment variables for server flags
+
+You can export the following environment variables based on the model you used.
+
+*   You can copy and export the `UNSCANNED_CKPT_PATH` from the model\_ckpt\_conversion.sh output.
```
Comment (yeandy) [inline README.md:102]:
```suggestion
*   You can copy and export the `UNSCANNED_CKPT_PATH` from the `model\_ckpt\_conversion.sh` output.
```
```suggestion
*   You can copy and export the `UNSCANNED_CKPT_PATH` from the model\_ckpt\_conversion.sh output.
```

**Comment by vipannalla** [comment]
Looks good, thanks!

**Review by vipannalla — APPROVED** [review]

File: README.md (hunk 4 above)
Comment (yeandy) [inline README.md:102]: Can you clarify this step? i.e. `UNSCANNED_CKPT_PATH` is a variable exported from `model_ckpt_conversion.sh`, which points to the unscanned checkpoint that it generated. 

And should we also explain that: Unscanned checkpoints are more efficient decoding, whereas scanned are better for training.

File: README.md [hunk 5]
```diff
@@ -4,6 +4,301 @@
...
+load_parameters_path=${LOAD_PARAMETERS_PATH} \
+max_prefill_predict_length=${MAX_PREFILL_PREDICT_LENGTH} \
+max_target_length=${MAX_TARGET_LENGTH} \
+model_name=${MODEL_NAME} \
+ici_fsdp_parallelism=${ICI_FSDP_PARALLELISM} \
+ici_autoregressive_parallelism=${ICI_AUTOREGRESSIVE_PARALLELISM} \
+ici_tensor_parallelism=${ICI_TENSOR_PARALLELISM} \
+scan_layers=${SCAN_LAYERS} \
+weight_dtype=${WEIGHT_DTYPE} \
+per_device_batch_size=${PER_DEVICE_BATCH_SIZE}
```
Comment (yeandy) [inline README.md:179]:
```suggestion
python MaxText/maxengine_server.py \
  MaxText/configs/base.yml \
  tokenizer_path=${TOKENIZER_PATH} \
  load_parameters_path=${LOAD_PARAMETERS_PATH} \
  max_prefill_predict_length=${MAX_PREFILL_PREDICT_LENGTH} \
  max_target_length=${MAX_TARGET_LENGTH} \
  model_name=${MODEL_NAME} \
  ici_fsdp_parallelism=${ICI_FSDP_PARALLELISM} \
  ici_autoregressive_parallelism=${ICI_AUTOREGRESSIVE_PARALLELISM} \
  ici_tensor_parallelism=${ICI_TENSOR_PARALLELISM} \
  scan_layers=${SCAN_LAYERS} \
  weight_dtype=${WEIGHT_DTYPE} \
  per_device_batch_size=${PER_DEVICE_BATCH_SIZE}
```

File: README.md [hunk 6]
```diff
@@ -4,6 +4,301 @@
...
+```
+
+### JetStream MaxText Server flag descriptions:
+
+
+
+*   tokenizer\_path: file path to a tokenizer (should match your model)
+*   load\_parameters\_path: Loads a just parameters from a specific directory
```
Comment (yeandy) [inline README.md:187]:
```suggestion
*   load\_parameters\_path: Loads the parameters (no optimizer states) from a specific directory
```

File: README.md [hunk 7]
```diff
@@ -4,6 +4,301 @@
...
+# For Gemma 7b, plz use Gemma tokenizer (must use a tokenizer that matches your model; double check if you are using tokenizer.gemma) and ShareGPT dataset; add `--warmup-first` flag for your 1st run to warmup the server.
+# Activate the python virtual environment we created in Step 2.
+cd ~
+source .env/bin/activate
+
+# download dataset
+wget https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/resolve/main/ShareGPT_V3_unfiltered_cleaned_split.json
+
+# run benchmark with the downloaded dataset and the tokenizer in maxtext
+# You can control the qps by setting `--request-rate`, the default value is inf.
```
Comment (yeandy) [inline README.md:258]: What is the effect of `--request-rate` being set to inf?

File: README.md [hunk 8]
```diff
@@ -4,6 +4,301 @@
...
+--dataset-path ~/ShareGPT_V3_unfiltered_cleaned_split.json \
+--max-output-length 1024 \
+--request-rate 5 \
+--warmup-first true
+```
+
+### Benchmarking Llama2-\*b
+
+```bash
+# Same as Gemma-7b except for tokenizer (must use a tokenizer that matches your model).
```
Comment (yeandy) [inline README.md:272]:
```suggestion
# Same as Gemma-7b except for the tokenizer (must use a tokenizer that matches your model, which should now be tokenizer.llama2). 
```

**Review by yeandy — APPROVED** [review]
A few suggestions, and questions, but overall LGTM!

File: README.md (hunk 2 above)
Comment (JoeZijunZhou) [reply README.md:37]: For this user guide demo, we want to use the stable release version. Later we can update it or keep it as it is.

File: README.md (hunk 7 above)
Comment (JoeZijunZhou) [reply README.md:258]: dispatch all requests at once.

File: README.md (hunk 2 above)
Comment (yeandy) [reply README.md:37]: Got it, thanks!

**Pull request merged**